In [10]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from pathlib import Path
import os
from sklearn.datasets import make_regression
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.ensemble import RandomForestClassifier

In [11]:
# Given DataFrames
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [12]:
# Read the 2019 training csv file into a pandas dataFrame
train_df = pd.read_csv(Path('Resources/2019loans.csv'))

# inspect the training DataFrame
pd.set_option('display.max_columns', 86)
train_df.head()

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0       57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1      141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2      321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3       11778   11778     3000.0    0.1240       100.22           RENT   
4      169382  169382    30000.0    0.1612      1056.49       MORTGAGE   

   annual_inc verification_status loan_status pymnt_plan    dti  delinq_2yrs  \
0    223000.0        Not Verified    low_risk          n  29.99          0.0   
1    123000.0     Source Verified    low_risk          n  11.26          2.0   
2    197000.0     Source Verified    low_risk          n  11.28          0.0   
3     45000.0        Not Verified    low_risk          n  18.08          0.0   
4    133000.0     Source Verified    low_risk          n  27.77          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0             0.0      15.0      0.0    39728.0       38.0   
1             0.0      16.0      0.0     9585.0       20.0   
2             0.0      12.0      0.0    16708.0       36.0   
3             0.0      12.0      1.0     8809.0       19.0   
4             2.0      13.0      0.0    65420.0       51.0   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w   12804.66       12804.66       953.33           953.33   
1                   w   15152.97       15152.97     10037.02         10037.02   
2                   w   15724.23       15724.23      6971.52          6971.52   
3                   w    2930.77        2930.77        81.63            81.63   
4                   w   20778.03       20778.03     14763.99         14763.99   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0           570.34         382.99                 0.0         0.0   
1          5847.03        4189.99                 0.0         0.0   
2          4275.77        2695.75                 0.0         0.0   
3            69.23          12.40                 0.0         0.0   
4          9221.97        5542.02                 0.0         0.0   

   collection_recovery_fee  last_pymnt_amnt  collections_12_mths_ex_med  \
0                      0.0           483.34                         0.0   
1                      0.0           478.68                         0.0   
2                      0.0           448.95                         0.0   
3                      0.0            85.76                         1.0   
4                      0.0          1056.49                         0.0   

   policy_code application_type  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0          1.0       Individual             0.0           0.0     465790.0   
1          1.0       Individual             0.0           0.0      60769.0   
2          1.0       Individual             0.0           0.0     617946.0   
3          1.0       Individual             0.0         568.0      20761.0   
4          1.0       Individual             0.0           0.0     237839.0   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0          2.0          7.0          3.0          7.0                 7.0   
1          0.0          2.0          0.0          1.0                17.0   
2          1.0          7.0          1.0          2.0                 5.0   
3          2.0          4.0          2.0          2.0                 8.0   
4          2.0          3.0          1.0          3.0                 5.0   

   total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  \
0       82290.0     70.0          1.0          1.0     24683.0      69.0   
1       18311.0     52.0          0.0          4.0      4311.0      50.0   
2       97335.0    108.0          0.0          0.0     11026.0      98.0   
3       11952.0     78.0          2.0          4.0      2409.0      49.0   
4       43636.0     74.0          1.0  

In [13]:
# Identify all the possible values in the loan_status column
list(train_df['loan_status'].unique())

['low_risk', 'high_risk']

In [14]:
# Convert categorical data to numeric and separate target feature for training data
y_train = train_df['loan_status']
X_train = train_df.drop(columns = ['loan_status'])
X_train.head()

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0       57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1      141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2      321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3       11778   11778     3000.0    0.1240       100.22           RENT   
4      169382  169382    30000.0    0.1612      1056.49       MORTGAGE   

   annual_inc verification_status pymnt_plan    dti  delinq_2yrs  \
0    223000.0        Not Verified          n  29.99          0.0   
1    123000.0     Source Verified          n  11.26          2.0   
2    197000.0     Source Verified          n  11.28          0.0   
3     45000.0        Not Verified          n  18.08          0.0   
4    133000.0     Source Verified          n  27.77          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0             0.0      15.0      0.0    39728.0       38.0   
1             0.0      16.0      0.0     9585.0       20.0   
2             0.0      12.0      0.0    16708.0       36.0   
3             0.0      12.0      1.0     8809.0       19.0   
4             2.0      13.0      0.0    65420.0       51.0   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w   12804.66       12804.66       953.33           953.33   
1                   w   15152.97       15152.97     10037.02         10037.02   
2                   w   15724.23       15724.23      6971.52          6971.52   
3                   w    2930.77        2930.77        81.63            81.63   
4                   w   20778.03       20778.03     14763.99         14763.99   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0           570.34         382.99                 0.0         0.0   
1          5847.03        4189.99                 0.0         0.0   
2          4275.77        2695.75                 0.0         0.0   
3            69.23          12.40                 0.0         0.0   
4          9221.97        5542.02                 0.0         0.0   

   collection_recovery_fee  last_pymnt_amnt  collections_12_mths_ex_med  \
0                      0.0           483.34                         0.0   
1                      0.0           478.68                         0.0   
2                      0.0           448.95                         0.0   
3                      0.0            85.76                         1.0   
4                      0.0          1056.49                         0.0   

   policy_code application_type  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0          1.0       Individual             0.0           0.0     465790.0   
1          1.0       Individual             0.0           0.0      60769.0   
2          1.0       Individual             0.0           0.0     617946.0   
3          1.0       Individual             0.0         568.0      20761.0   
4          1.0       Individual             0.0           0.0     237839.0   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0          2.0          7.0          3.0          7.0                 7.0   
1          0.0          2.0          0.0          1.0                17.0   
2          1.0          7.0          1.0          2.0                 5.0   
3          2.0          4.0          2.0          2.0                 8.0   
4          2.0          3.0          1.0          3.0                 5.0   

   total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  \
0       82290.0     70.0          1.0          1.0     24683.0      69.0   
1       18311.0     52.0          0.0          4.0      4311.0      50.0   
2       97335.0    108.0          0.0          0.0     11026.0      98.0   
3       11952.0     78.0          2.0          4.0      2409.0      49.0   
4       43636.0     74.0          1.0          1.0     19804.0      74.0   

   total_rev_hi_lim  inq_fi  total

In [15]:
# Convert categorical data to numeric and separate target feature for testing data
y_test = test_df['loan_status']
X_test = test_df.drop(columns = ['loan_status'])
X_test.head()

Unnamed: 0  index  loan_amnt  int_rate  installment home_ownership  \
0       67991  67991    40000.0    0.0819       814.70       MORTGAGE   
1       25429  25429     6000.0    0.1524       208.70           RENT   
2       38496  38496     3600.0    0.1695       128.27           RENT   
3       19667  19667    20000.0    0.1524       478.33           RENT   
4       37505  37505     3600.0    0.1240       120.27           RENT   

   annual_inc verification_status pymnt_plan    dti  delinq_2yrs  \
0    140000.0        Not Verified          n  19.75          0.0   
1     55000.0        Not Verified          n  11.52          2.0   
2     42000.0        Not Verified          n   6.74          0.0   
3    100000.0        Not Verified          n  12.13          0.0   
4     50000.0        Not Verified          n  16.08          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0             1.0      18.0      0.0     9471.0       45.0   
1             0.0       8.0      0.0     1280.0       12.0   
2             0.0       6.0      0.0     4757.0        6.0   
3             2.0       7.0      0.0    12731.0       21.0   
4             3.0       6.0      0.0    10413.0        6.0   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w   33249.95       33249.95      9740.00          9740.00   
1                   w       0.00           0.00      6185.59          6185.59   
2                   w       0.00           0.00      3603.73          3603.73   
3                   w   17624.06       17624.06      4766.37          4766.37   
4                   w       0.00           0.00      3602.48          3602.48   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          6750.05        2989.95                 0.0         0.0   
1          6000.00         185.59                 0.0         0.0   
2          3600.00           3.73                 0.0         0.0   
3          2375.94        2390.43                 0.0         0.0   
4          3600.00           2.48                 0.0         0.0   

   collection_recovery_fee  last_pymnt_amnt  collections_12_mths_ex_med  \
0                      0.0           814.70                         0.0   
1                      0.0          5773.27                         0.0   
2                      0.0          3607.12                         0.0   
3                      0.0           478.33                         0.0   
4                      0.0          3604.96                         0.0   

   policy_code application_type  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0          1.0       Individual             0.0           0.0     377773.0   
1          1.0       Individual             0.0           0.0      23460.0   
2          1.0       Individual             0.0           0.0      19183.0   
3          1.0       Individual             0.0           0.0      43817.0   
4          1.0       Individual             0.0           0.0      32448.0   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0          1.0          6.0          1.0          2.0                 9.0   
1          1.0          1.0          1.0          1.0                 6.0   
2          0.0          3.0          0.0          0.0                40.0   
3          1.0          2.0          1.0          3.0                11.0   
4          2.0          2.0          1.0          1.0                 9.0   

   total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  \
0       61443.0     51.0          1.0          1.0      3218.0      26.0   
1       22180.0     94.0          0.0          1.0       713.0      68.0   
2       14426.0     96.0          1.0          2.0      4757.0      83.0   
3       31086.0     86.0          3.0          3.0      5917.0      78.0   
4       22035.0     88.0          2.0          2.0      5985.0      71.0   

   total_rev_hi_lim  inq_fi  total_cu_tl

In [16]:
# Get dummy varaibles for both the testing and training data 
X_dummies_train = pd.get_dummies(X_train)
X_dummies_test = pd.get_dummies(X_test)

print(f"Training set{X_dummies_train.shape}, Test set{X_dummies_test.shape}")

Training set(12180, 94), Test set(4702, 93)


In [17]:
# Convert output labels to 0 and 1 for training
y_label_train = LabelEncoder().fit_transform(train_df['loan_status'])
y_label_train

# Convert output labels to 0 and 1 for testing
y_label_test = LabelEncoder().fit_transform(test_df['loan_status'])
y_label_test

array([1, 1, 1, ..., 0, 0, 0])

In [18]:
# appened missing dummy variables from training to testing because there are categoreies that don't exist in the testing set
for column in X_dummies_train.columns:
    if column not in X_dummies_test.columns:
        X_dummies_test[column] = 0

## Consideration hypothesis

I think that the random forest classifier will do a better job than the logistic regression model, because a decision tree irrerates X number of times and averages the output, but ech one of those individual outputs is attempting to make the best possible solution, and then each one of those solutions is randomly sampling the dataset by RNGing the inputs, but they aren't over represeneted if an instance occured multiple times because, you know, randomness.

In [19]:
# check impact
print(f"Training set{X_dummies_train.shape}, Test set{X_dummies_test.shape}")

Training set(12180, 94), Test set(4702, 94)


In [20]:
# Train the Logistic Regression model on the unscaled data and print the model score
clf_logistic  = LogisticRegression().fit(X_dummies_train, y_label_train)

print(f"Logistic Regression(unscaled) Training Score {clf_logistic.score(X_dummies_train, y_label_train)}.")

Logistic Regression(unscaled) Training Score 0.6485221674876848.


C:\Users\mdebo\anaconda3\envs\PythonData\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [21]:
claf_logistic  = LogisticRegression().fit(X_dummies_test, y_label_test)

print(f"Logistic Regression (unscaled) Testing Score {clf_logistic.score(X_dummies_test, y_label_test)}.")

Logistic Regression (unscaled) Testing Score 0.5253083794130158.


C:\Users\mdebo\anaconda3\envs\PythonData\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [22]:
# Train the Random Forest model on the unscaled data and print the model score
clf_forest = RandomForestClassifier(random_state=42, n_estimators=250).fit(X_dummies_train, y_label_train)

print(f"Random Forest (unscaled) Training Score: {clf_forest.score(X_dummies_train, y_label_train)}.")

Random Forest (unscaled) Training Score: 1.0.


In [23]:
clf_forest = RandomForestClassifier(random_state=42, n_estimators=200).fit(X_dummies_train, y_label_train)

print(f"Random Forest (unscaled) Testing Score: {clf_forest.score(X_dummies_test, y_label_test)}.")

Random Forest (unscaled) Testing Score: 0.623139089749043.


## Consideration hypothesis the second

Okay so on the unscaled tests, the training scores of the LR and RFC is where one expects it on the LR and something is off on the RFC because I don't care how good it is, it shouldn't be 100%. Were not comparing it to itself, and when considering the testing scores, yes there is the drop-off of improvement, but it actually seems like the Logistic Regression did better because its a drop in accuracy by 12% but when compared to the drop in accuracy of the random forest classifier, which dropped by 37%, comparatively it seems like it is doing better.

In [24]:
# Scale the data

scaler = StandardScaler().fit(X_dummies_train)
X_train_scaled = scaler.transform(X_dummies_train)
X_test_scaled = scaler.transform(X_dummies_test)

In [25]:
# Train the Logistic Regression model on the scaled data and print the model score 
clf_lr = LogisticRegression().fit(X_train_scaled, y_label_train)

print(f'Logistic Regression (scaled) Training Score: {clf_lr.score(X_train_scaled, y_label_train)}')

Logistic Regression (scaled) Training Score: 0.713136288998358


C:\Users\mdebo\anaconda3\envs\PythonData\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [26]:
clf_lr = LogisticRegression().fit(X_test_scaled, y_label_test)

print(f'Logistic Regression (scaled) Testing Score: {clf_lr.score(X_test_scaled, y_label_test)}')

Logistic Regression (scaled) Testing Score: 0.8928115695448745


C:\Users\mdebo\anaconda3\envs\PythonData\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [27]:
# Train a Random Forest Classifier model on the scaled data and print the model score
clf_rfc = RandomForestClassifier(random_state=42, n_estimators=500).fit(X_train_scaled, y_label_train)

print(f'Random Forest Training Score: {clf_rfc.score(X_train_scaled, y_label_train)}')

Random Forest Training Score: 1.0


In [28]:
clf_rfc = RandomForestClassifier(random_state=42, n_estimators=500).fit(X_test_scaled, y_label_test)

print(f'Random Forest Testing Score: {clf_rfc.score(X_test_scaled, y_label_test)}')

Random Forest Testing Score: 1.0


In [29]:
# Create the Confusion Matrix on scaled Logistic Regression Model

y_true_lr = y_label_test
y_pred_lr = clf_lr.predict(X_test_scaled)
confusion_matrix(y_true_lr, y_pred_lr)

array([[1993,  358],
       [ 146, 2205]], dtype=int64)

In [30]:
tn, fp, fn, tp = confusion_matrix(y_true_lr, y_pred_lr).ravel()
accuracy = (tp + tn) / (tp + fp + tn + fn)
print(f"Accuracy: {accuracy}")

Accuracy: 0.8928115695448745


In [31]:
# Reshape for specifc
matrix = confusion_matrix(y_true_lr, y_pred_lr)
matrix = matrix.astype('float') / matrix.sum(axis=1)[:, np.newaxis]

In [32]:
# Classification Report on scaled Logistic Regression model

target_names = ['low_risk', 'high_risk']
print(classification_report(y_true_lr, y_pred_lr, target_names=target_names))

              precision    recall  f1-score   support

    low_risk       0.93      0.85      0.89      2351
   high_risk       0.86      0.94      0.90      2351

    accuracy                           0.89      4702
   macro avg       0.90      0.89      0.89      4702
weighted avg       0.90      0.89      0.89      4702



In [33]:
# Create the Confusion Matrix on scaled Random Forest Classifier model

y_true_rfc = y_label_test
y_pred_rfc = clf_rfc.predict(X_test_scaled)
confusion_matrix(y_true_rfc, y_pred_rfc)

array([[2351,    0],
       [   0, 2351]], dtype=int64)

In [34]:
tn, fp, fn, tp = confusion_matrix(y_true_rfc, y_pred_rfc).ravel()
accuracy = (tp + tn) / (tp + fp + tn + fn)
print(f"Accuracy: {accuracy}")

Accuracy: 1.0


In [35]:
# Reshape for specifc
matrix = confusion_matrix(y_true_rfc, y_pred_rfc)
matrix = matrix.astype('float') / matrix.sum(axis=1)[:, np.newaxis]

In [36]:
# Classification Report on scaled Random Forest Classifier model

target_names = ['low_risk', 'high_risk']
print(classification_report(y_true_rfc, y_pred_rfc, target_names=target_names))

              precision    recall  f1-score   support

    low_risk       1.00      1.00      1.00      2351
   high_risk       1.00      1.00      1.00      2351

    accuracy                           1.00      4702
   macro avg       1.00      1.00      1.00      4702
weighted avg       1.00      1.00      1.00      4702



## Hypothesis testing

So in review, the random forest was not better, but I think that something is off and I will have to go back and check. Its either the case that when comparing training data to itself should always return 100% accuracy, seen here in the training score of 1.0 of the RFC but NOT OF the LR model which was like 0.62, and this is on the unscailed. But that didn't play out, and so something was off in the beginning I think with the RFC, and this becomes magnified in the scaled portion. Again, here the LR just does a 'better' job, it improved at each step of this analysis, and its not great, each of the improvements but they are marginally better by some like some 10% increase each time. I included the confusion matrix and classifcation report just to check myself, and then went back over them with the accuracy printed value to double check myself and there is something up with the random forest from the beginning. It was either the random_states being set at 42 in the beginnging, and that clouded it from that point forward, or I messed something up that someone will udoubtedly point out to me. 

But at the end of the day it was the logistic regression model that was better, its improved with scaling, and it has a believable breakdown on the accuracy, confusion matrix and classification report.